pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,430 @@
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(err, "Failed to send ICMP message to Node", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Failed to send ICMP message to Node")
+		}
+	}
+}
+
+// Run starts the NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+// monitorLoop is the main loop to monitor the latency of the Node.
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-stopCh:
+			// Close current sockets, the recvPing will stop by itself
+			if ipv4Socket != nil {
+				ipv4Socket.Close()
+			}
+			if ipv6Socket != nil {
+				ipv6Socket.Close()
+			}
+			// Stop the ticker loop
+			if ticker != nil {
+				ticker.Stop()
+			}
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// latencyConfig changed
+				updateTicker(m.latencyConfig.Interval)
+
+				// If the recvPing socket is closed, restart it
+				// In case of IPv4-only or IPv6-only, we need to check the socket status,
+				// and restart it if it is closed(CRD is deleted).
+				if ipv4Socket == nil && m.isIPv4Enabled {
+					// Create a new socket for IPv4 when the gatewayConfig is IPv4-only
+					ipv4Socket, err = icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
+					if err != nil {
+						klog.ErrorS(err, "Failed to create ICMP socket for IPv4")
+						return
+					}
+					go m.recvPing(ipv4Socket, true)
+				}
+				if ipv6Socket == nil && m.isIPv6Enabled {
+					// Create a new socket for IPv6 when the gatewayConfig is IPv6-only
+					ipv6Socket, err = icmp.ListenPacket(IPv6ProtocolICMPRaw, "::")
+					if err != nil {
+						klog.ErrorS(err, "Failed to create ICMP socket for IPv6")
+						return
+					}
+					go m.recvPing(ipv6Socket, false)
+				}
+			} else {
+				// latencyConfig deleted
+				if ticker != nil {
+					ticker.Stop()
+					ticker = nil
+				}
+				m.latencyStore.CleanUp()
```

Comment:
Shouldn't we close the sockets as well, and set them to `nil`? I don't think we want to keep `recvPing` running if the latency config is deleted / disabled.

It is important to close them _after_ stopping the ticker

Reply:
That's right, I have added the `close` method after the ticker is stopped.